### ISSUE TITLE
Building for linux is broken beyond repair.

### ISSUE THREAD

**Issue by jsmith111** [aseprite/aseprite#574]
First of all, simpleini and the pixman library are COMPLETELY MISSING from the entire source directory, they are not there, they are non-existant.

So after painstakingly finding these from some other websites and manually placing them where they should be cmake finally succeeds, and after running make you get this:
```
~/Documents/aseprite-master/build$ make
[ 11%] Built target libcurl
[ 11%] Built target scripting-lib
[ 12%] Built target she
[ 18%] Built target jpeg
[ 19%] Built target updater-lib
[ 19%] Built target net-lib
[ 21%] Built target zlib
[ 24%] Built target base-lib
[ 36%] Built target allegro
[ 36%] Built target cfg-lib
[ 37%] Built target css-lib
[ 41%] Built target doc-lib
[ 41%] Building CXX object src/render/CMakeFiles/render-lib.dir/quantization.cpp.o
In file included from /home/null/Documents/aseprite-master/src/./render/color_histogram.h:18:0,
                 from /home/null/Documents/aseprite-master/src/./render/quantization.h:15,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:11:
/home/null/Documents/aseprite-master/src/./render/median_cut.h: In member function ‘uint32_t render::Box<Histogram>::meanColor(const Histogram&) const’:
/home/null/Documents/aseprite-master/src/./render/median_cut.h:100:33: error: there are no arguments to ‘rgba’ that depend on a template parameter, so a declaration of ‘rgba’ must be available [-fpermissive]
         return rgba(0, 0, 0, 255);
                                 ^
/home/null/Documents/aseprite-master/src/./render/median_cut.h:100:33: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
In file included from /home/null/Documents/aseprite-master/src/./doc/image_bits.h:12:0,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:15:
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h: In member function ‘bool doc::BitPixelAccess::operator==(int) const’:
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h:212:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       return (color_t(*this) == b);
                                 ^
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h: In member function ‘bool doc::BitPixelAccess::operator!=(int) const’:
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h:224:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       return (color_t(*this) != b);
                                 ^
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h: In function ‘bool doc::operator==(int, const doc::BitPixelAccess&)’:
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h:244:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     return (a == color_t(b));
                           ^
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h: In function ‘bool doc::operator!=(int, const doc::BitPixelAccess&)’:
/home/null/Documents/aseprite-master/src/./doc/image_iterator.h:252:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     return (a != color_t(b));
                           ^
In file included from /home/null/Documents/aseprite-master/src/./render/color_histogram.h:18:0,
                 from /home/null/Documents/aseprite-master/src/./render/quantization.h:15,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:11:
/home/null/Documents/aseprite-master/src/./render/median_cut.h: In instantiation of ‘uint32_t render::Box<Histogram>::meanColor(const Histogram&) const [with Histogram = render::ColorHistogram<5, 6, 5>; uint32_t = unsigned int]’:
/home/null/Documents/aseprite-master/src/./render/median_cut.h:274:11:   required from ‘void render::median_cut(const Histogram&, size_t, std::vector<unsigned int>&) [with Histogram = render::ColorHistogram<5, 6, 5>; size_t = long unsigned int]’
/home/null/Documents/aseprite-master/src/./render/color_histogram.h:95:44:   required from ‘int render::ColorHistogram<RBits, GBits, BBits>::createOptimizedPalette(doc::Palette*, int, int) [with int RBits = 5; int GBits = 6; int BBits = 5]’
/home/null/Documents/aseprite-master/src/render/quantization.cpp:479:70:   required from here
/home/null/Documents/aseprite-master/src/./render/median_cut.h:100:33: error: ‘rgba’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
         return rgba(0, 0, 0, 255);
                                 ^
In file included from /home/null/Documents/aseprite-master/src/./doc/image.h:12:0,
                 from /home/null/Documents/aseprite-master/src/./render/color_histogram.h:14,
                 from /home/null/Documents/aseprite-master/src/./render/quantization.h:15,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:11:
/home/null/Documents/aseprite-master/src/./doc/color.h:47:19: note: ‘uint32_t doc::rgba(uint8_t, uint8_t, uint8_t, uint8_t)’ declared here, later in the translation unit
   inline uint32_t rgba(uint8_t r, uint8_t g, uint8_t b, uint8_t a) {
                   ^
In file included from /home/null/Documents/aseprite-master/src/./render/color_histogram.h:18:0,
                 from /home/null/Documents/aseprite-master/src/./render/quantization.h:15,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:11:
/home/null/Documents/aseprite-master/src/./render/median_cut.h:105:68: error: ‘rgba’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
                   (255 * b / (Histogram::BElements-1)) / count, 255);
                                                                    ^
In file included from /home/null/Documents/aseprite-master/src/./doc/image.h:12:0,
                 from /home/null/Documents/aseprite-master/src/./render/color_histogram.h:14,
                 from /home/null/Documents/aseprite-master/src/./render/quantization.h:15,
                 from /home/null/Documents/aseprite-master/src/render/quantization.cpp:11:
/home/null/Documents/aseprite-master/src/./doc/color.h:47:19: note: ‘uint32_t doc::rgba(uint8_t, uint8_t, uint8_t, uint8_t)’ declared here, later in the translation unit
   inline uint32_t rgba(uint8_t r, uint8_t g, uint8_t b, uint8_t a) {
                   ^
make[2]: *** [src/render/CMakeFiles/render-lib.dir/quantization.cpp.o] Error 1
make[1]: *** [src/render/CMakeFiles/render-lib.dir/all] Error 2
make: *** [all] Error 2
~/Documents/aseprite-master/build$ 
```

I'm not really a developer at all, I don't know anything, but I fixed this myself by changing this code in src/render/median_cut.h from this:
```
      // No colors in the box? This should not be possible.
      ASSERT(count > 0 && "Box without histogram points, you must fill the histogram before using this function.");
      if (count == 0)
        return rgba(0, 0, 0, 255);

      // Returns the mean.
      return rgba((255 * r / (Histogram::RElements-1)) / count,
                  (255 * g / (Histogram::GElements-1)) / count,
                  (255 * b / (Histogram::BElements-1)) / count, 255);
```

to this:
```
      // No colors in the box? This should not be possible.
      ASSERT(count > 0 && "Box without histogram points, you must fill the histogram before using this function.");
      if (count == 0)
        return doc::rgba(0, 0, 0, 255);

      // Returns the mean.
      return doc::rgba((255 * r / (Histogram::RElements-1)) / count,
                  (255 * g / (Histogram::GElements-1)) / count,
                  (255 * b / (Histogram::BElements-1)) / count, 255);
```

Now, I'm at this point getting this error:
```
~/Documents/aseprite-master/build$ make
[ 11%] Built target libcurl
[ 11%] Built target scripting-lib
[ 12%] Built target she
[ 18%] Built target jpeg
[ 19%] Built target updater-lib
[ 19%] Built target net-lib
[ 21%] Built target zlib
[ 24%] Built target base-lib
[ 36%] Built target allegro
[ 36%] Built target cfg-lib
[ 37%] Built target css-lib
[ 41%] Built target doc-lib
[ 41%] Built target render-lib
[ 42%] Built target filters-lib
[ 42%] Built target fixmath-lib
[ 43%] Built target gfx-lib
[ 48%] Built target ui-lib
[ 48%] Built target undo-lib
[ 49%] Built target tinyxml
[ 49%] Building CXX object src/gen/CMakeFiles/gen.dir/pref_types.cpp.o
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp: In function ‘void print_pref_class_def(TiXmlElement*, const string&, const char*, int)’:
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:41:46: error: ‘runtime_error’ is not a member of ‘std’
         if (!child->Attribute("type")) throw std::runtime_error("missing 'type' attr in <option>");
                                              ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:42:29: error: ‘runtime_error’ is not a member of ‘std’
         if (!childId) throw std::runtime_error("missing 'id' attr in <option>");
                             ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:48:29: error: ‘runtime_error’ is not a member of ‘std’
         if (!childId) throw std::runtime_error("missing 'id' attr in <section>");
                             ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp: In function ‘void print_pref_class_impl(TiXmlElement*, const string&, const string&, const char*)’:
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:81:46: error: ‘runtime_error’ is not a member of ‘std’
         if (!child->Attribute("type")) throw std::runtime_error("missing 'type' attr in <option>");
                                              ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:82:29: error: ‘runtime_error’ is not a member of ‘std’
         if (!childId) throw std::runtime_error("missing 'id' attr in <option>");
                             ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:91:29: error: ‘runtime_error’ is not a member of ‘std’
         if (!childId) throw std::runtime_error("missing 'id' attr in <option>");
                             ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp: In function ‘void gen_pref_header(TiXmlDocument*, const string&)’:
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:202:39: error: ‘runtime_error’ is not a member of ‘std’
     if (!elem->Attribute("id")) throw std::runtime_error("missing 'id' attr in <enum>");
                                       ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:209:42: error: ‘runtime_error’ is not a member of ‘std’
       if (!child->Attribute("id")) throw std::runtime_error("missing 'id' attr in <value>");
                                          ^
/home/null/Documents/aseprite-master/src/gen/pref_types.cpp:210:45: error: ‘runtime_error’ is not a member of ‘std’
       if (!child->Attribute("value")) throw std::runtime_error("missing 'value' attr in <value>");
                                             ^
make[2]: *** [src/gen/CMakeFiles/gen.dir/pref_types.cpp.o] Error 1
make[1]: *** [src/gen/CMakeFiles/gen.dir/all] Error 2
make: *** [all] Error 2
~/Documents/aseprite-master/build$ 
```
I have not yet fixed this error.

Building aseprite for linux is extremely painful and soul crushing.

**Comment by jsmith111** [comment]
Ok, so the last error from the previous was fixed by changing line 11 in /src/base/exception.h from:
```
#include <exception>
```

to:
```
#include <stdexcept>
```

And here is more than likely, maybe, hopefully the final error that I haven't yet solved:
```
~/Documents/aseprite-master/build$ make
[ 11%] Built target libcurl
[ 11%] Built target scripting-lib
[ 12%] Built target she
[ 18%] Built target jpeg
[ 19%] Built target updater-lib
[ 19%] Built target net-lib
[ 21%] Built target zlib
[ 24%] Built target base-lib
[ 36%] Built target allegro
[ 36%] Built target cfg-lib
[ 37%] Built target css-lib
[ 41%] Built target doc-lib
[ 41%] Built target render-lib
[ 42%] Built target filters-lib
[ 42%] Built target fixmath-lib
[ 43%] Built target gfx-lib
[ 48%] Built target ui-lib
[ 48%] Built target undo-lib
[ 49%] Built target tinyxml
[ 49%] Built target gen
[ 82%] Built target app-lib
[ 90%] Built target copy_data
[ 91%] Built target libpng
[ 92%] Built target giflib
[ 96%] Built target pixman
[ 97%] Built target loadpng
Linking CXX executable ../bin/aseprite
/usr/bin/ld: cannot find -lsimpleini
collect2: error: ld returned 1 exit status
make[2]: *** [bin/aseprite] Error 1
make[1]: *** [src/CMakeFiles/aseprite.dir/all] Error 2
make: *** [all] Error 2
~/Documents/aseprite-master/build$ 
```

**Comment by jsmith111** [comment]
Ok, everything finally compiled, to fix the last thing I had to run through every link.txt file in each directory in /build/src/Cmakefiles/ and remove:
```
-lsimpleini
```
from each one

which isn't the correct solution.

I ran aseprite and it seems to be working properly, and what's more is there is no mouse lag really, which was present last time I tried aseprite.

**Comment by dacap** [comment]
Hi there! Today I've merged two pull requests to fix these problems: #571 and #572. About the source code, at the moment the only way to get the full source code is using Git and submodules. The process is described in the [CONTRIBUTING file](https://github.com/aseprite/aseprite/blob/master/CONTRIBUTING.md#get-the-source-code)
